sql/parser/builtins.go

```diff
@@ -72,6 +74,67 @@ var builtins = map[string]builtin{
 	"upper": stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToUpper(s)), nil
 	}),
+
+	// TODO(XisiHuang): support the substring(str FROM x FOR y) syntax
+	"substring": {
+		nArgs: -1,
+		fn: func(args DTuple) (Datum, error) {
+			argsNum := len(args)
+			if argsNum != 2 && argsNum != 3 {
+				return DNull, fmt.Errorf("incorrect number of arguments: %d vs %s",
+					argsNum, "2 or 3")
+			}
+
+			dstr, ok := args[0].(DString)
+			if !ok {
+				return DNull, argTypeError(args[0], "string")
+			}
+			str := string(dstr)
+
+			start, ok := args[1].(DInt)
+			if !ok {
+				return DNull, argTypeError(args[1], "int")
+			}
+
+			s := int(start)
+			if s < 0 || s > len(str)-1 {
```

Comment:
I'm not sure why, but its in the spec: https://golang.org/ref/spec#Slice_expressions

```
For arrays or strings, the indices are in range if 0 <= low <= high <= len(a), 
otherwise they are out of range.
```